### PR TITLE
CORDA-2050 Upgrade Corda to Java 11 (compatibility mode)

### PR DESCRIPTION
**Pull request by josecoll**
## Summary
This PR contains the first phase migration of Corda to build and execute on a Java 11 VM.
The codebase maintains 100% source code compatibility with Java 8 (ie. changes do not use any of the new Java 9+ JDK libraries). This compatibility mode entails all code executes within the global `UNNAMED` module space (eg. the equivalent of the Java 8 global classpath), so there is no effective true module isolation per se.

We use the Gradle `sourceCompatibility` and `targetCompatibility` flags to specify compatibility modes, such that:
-  `sourceCompatibility = VERSION_1_8`
-  `targetCompatibility = VERSION_1_8` or `VERSION_11` (dependent on JDK used to execute gradle)

**Note**: a specific patched version of Java 11 is being used due to an officially reported OpenJDK bug related to cryptographic libraries used by Corda. The official OpenJDK 11 patch is expected on 15th October 2019.

Please see [JDK 11 Migration](https://r3-cev.atlassian.net/wiki/spaces/CCD/pages/1183909338/JDK+11+Migration) for further background and detailed migration steps completed.

## Dependencies
The Java 11 build of Corda requires:
- an upgraded version of Quasar built for Java 11.
  The version of Quasar to use may be driven from system properties (the default is to use the `quasarVersion` specified in `constants.properties`)
   For Java 11, these must be set as follows: `-DquasarVersion=0.8.0 -DquasarClassifier`
- Corda Gradle plugins upgraded to use aforementioned version of Quasar - see [CORDA-2924](https://r3-cev.atlassian.net/browse/CORDA-2924)
- Gradle 5: existing master codebase has already been upgraded to use this - see [CORDA-2893](https://r3-cev.atlassian.net/browse/CORDA-2893) (and an associated task to upgrade the capsule plugin to use Gradle 5: [CORDA-2895](https://r3-cev.atlassian.net/browse/CORDA-2895)).
- **a patched version of the official OpenJDK 11** (note: we initially patched the OpenJDK in-house, subsequently reported and obtained an official patch from our JDK supplier, and are now awaiting an official patch due on or around the 15th October 2019 in line with [Oracle's Critical Patch Update](https://www.oracle.com/technetwork/topics/security/alerts-086861.html) quarterly release). The OpenJDK reported bug in question can be found [here](https://bugs.openjdk.java.net/browse/JDK-8225745).

## Constraints
The Kotlin compiler does not support usage of Java 9 module directives (eg. to specify that private internal JDK packages may still be used in the global UNNAMED module space). It was necessary to re-write the `SharedMemoryIncremental` Kotlin class in Java to be able to specify the use of a private internal package at compile time as follows:
```
        if (JavaVersion.current() == JavaVersion.VERSION_11)
            options.compilerArgs = [
                    '--add-exports', 'java.base/sun.nio.ch=ALL-UNNAMED'
            ]
```

Corda Java 11 docker images included in this PR (eg. see `Dockerfile11`) temporarily use an in-house dockerized patched JDK 11 image published to an internal R3 docker registry.

## DJVM and SGX compatibility
The modules `core`, `core-deterministic`, `serialization`, `serialization-deterministic` strictly target Java 8 JVM bytecode to support deployment and execution of contract verification logic in an SGX hosted deterministic JVM.

## Testing
- All Unit, Integration and Smoke tests have been executed in a Continuous Integration environment (as per Corda master).
- Basic samples (Bank Of Corda) and compatibility tests (eg. a transacting Behave scenario combining Corda nodes built with Java 8 and 11).
- Basic performance testing using the R3 performance cluster (no specific Java 11 JVM / GC tuning applied)
- Specific tools and samples require additional testing.

## Outstanding issues
- Node Explorer does not currently work due to code requiring upgrade to latest stable version of  [TornadoFX](https://github.com/edvin/tornadofx/tree/jdk10) (and then switchover to a SNAPSHOT version with Java 10+ support). Work to complete this is being tracked here: https://r3-cev.atlassian.net/browse/CORDA-3208

## Further work
This PR is part of a larger Epic described [here](https://r3-cev.atlassian.net/browse/CID-725), and which includes future activities such as:
- true (Java 9) modularisation (inclusion of `module-info.java` and explicit imports/exports)
- adoption of new Java 11 JDK libraries and language features - [CORDA-2051](https://r3-cev.atlassian.net/browse/CORDA-2051)
- use JLink to create optimised runtime images - [CORDA-2754](https://r3-cev.atlassian.net/browse/CORDA-2754)
- improve startup and footprint using Application Class Data Sharing (AppCDS) - [CORDA-2755](https://r3-cev.atlassian.net/browse/CORDA-2755)

### CODE REVIEW

**Comment by mikehearn** [comment]
171 commits .... should probably be squashed a bit! :o

File: build.gradle [hunk 1]
```diff
@@ -14,7 +17,13 @@ buildscript {
     ext.kotlin_version = constants.getProperty("kotlinVersion")
 
     ext.quasar_group = 'co.paralleluniverse'
-    ext.quasar_version = constants.getProperty("quasarVersion")
+    // Java 11: pass in -DquasarVersion=0.8.0 -DquasarClassifier
```
Comment (mikehearn) [inline build.gradle]: Can we set these automatically based on the version of Java, as seen below?

File: client/jfx/src/main/java/net/corda/client/jfx/utils/ReplayedList.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package net.corda.client.jfx.utils;
+
+import javafx.collections.ObservableList;
+import org.jetbrains.annotations.NotNull;
+
+// Java 9 introduces a new abstract method that we need to override (without using the explicit Kotlin `override` keyword to be backwards compatible
```
Comment (mikehearn) [inline client/jfx/src/main/java/net/corda/client/jfx/utils/ReplayedList.java:6]: Hm, not sure I fully understand this. What's wrong with using the override keyword? Wouldn't work with Java 8?

I wonder if it's worth being compatible with both 8 and 11 simultaneously - that's the source of a lot of complexity here. What do we lose if we just take the leap and use git rollbacks to go to back 8 if we need to?

File: client/rpc/build.gradle [hunk 3]
```diff
@@ -98,6 +98,9 @@ dependencies {
     smokeTestImplementation "junit:junit:$junit_version"
     smokeTestRuntimeOnly "org.junit.vintage:junit-vintage-engine:${junit_vintage_version}"
     smokeTestRuntimeOnly "org.junit.platform:junit-platform-launcher:${junit_platform_version}"
+
+    // JDK11: required by Quasar at run-time
```
Comment (mikehearn) [inline client/rpc/build.gradle:102]: Huh, why is Quasar getting involved in the RPC client code? We are only meant to use it on the server.

File: node/src/main/kotlin/net/corda/node/serialization/kryo/DefaultKryoCustomizer.kt [hunk 4]
```diff
@@ -97,7 +95,6 @@ object DefaultKryoCustomizer {
             register(BufferedInputStream::class.java, InputStreamSerializer)
             register(Class.forName("sun.net.www.protocol.jar.JarURLConnection\$JarURLInputStream"), InputStreamSerializer)
             noReferencesWithin<WireTransaction>()
-            register(ECPublicKeyImpl::class.java, publicKeySerializer)
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/serialization/kryo/DefaultKryoCustomizer.kt:100]: Hm, why are these removed?

File: node/src/main/kotlin/net/corda/node/utilities/JVMAgentUtil.kt [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package net.corda.node.utilities
+
+import com.sun.tools.attach.VirtualMachine
+import org.slf4j.Logger
+import java.lang.management.ManagementFactory
+import java.util.*
+
+object JVMAgentUtil {
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/utilities/JVMAgentUtil.kt:8]: I wonder if we really need this anymore.

File: samples/simm-valuation-demo/build.gradle [hunk 6]
```diff
@@ -31,14 +31,21 @@ dependencies {
     // The SIMM demo CorDapp depends upon Cash CorDapp features
     cordapp project(':finance:contracts')
     cordapp project(':finance:workflows')
-    cordapp project(path: ':samples:simm-valuation-demo:contracts-states', configuration: 'shrinkArtifacts')
+    cordapp project(':samples:simm-valuation-demo:contracts-states')
     cordapp project(':samples:simm-valuation-demo:flows')
 
+    // Cordformation needs a SLF4J implementation when executing the Network
```
Comment (mikehearn) [inline samples/simm-valuation-demo/build.gradle:37]: Pretty grim that we're complicating the build file for apps even further. Developer experience is going down the toilet - this magic bit of boilerplate isn't documented anywhere. Is there no fix on the log4j side, or can we hide this away in the Gradle plugins somehow?

File: samples/simm-valuation-demo/contracts-states/build.gradle [hunk 7]
```diff
@@ -8,7 +8,7 @@ cordapp {
     minimumPlatformVersion 1
     signing {
         // Cordapp is signed after the "shrink" task.
-        enabled false
+        enabled true
```
Comment (mikehearn) [inline samples/simm-valuation-demo/contracts-states/build.gradle]: ?

File: serialization/src/main/kotlin/net/corda/serialization/internal/carpenter/ClassCarpenter.kt [hunk 8]
```diff
@@ -346,7 +346,7 @@ class ClassCarpenterImpl @JvmOverloads constructor (override val whitelist: Clas
             visitCode()
             visitLdcInsn(Type.getType("L${schema.jvmName};"))
             visitVarInsn(ALOAD, 0)
-            visitMethodInsn(INVOKESTATIC, jlEnum, "valueOf", "(L$jlClass;L$jlString;)L$jlEnum;", true)
+            visitMethodInsn(INVOKESTATIC, jlEnum, "valueOf", "(L$jlClass;L$jlString;)L$jlEnum;", false)
```
Comment (mikehearn) [inline serialization/src/main/kotlin/net/corda/serialization/internal/carpenter/ClassCarpenter.kt:349]: Wasn't this previously fixed to be correct? IIRC this flag is controlling whether it's an interface method or something like that, why is it changing here?

**Review by mikehearn — COMMENTED** [review]
Wow, great work. Truly a mission.

How is packageName implemented in Java 11? I wonder if the packageName_ change should just be cleaned up so we use the real support in Java 11 (or package.name) and undo the nasty optimisations, knowing we'll get the speed back as part of the upgrade?

File: client/jfx/src/main/java/net/corda/client/jfx/utils/ReplayedList.java (hunk 2 above)
Comment (josecoll) [reply client/jfx/src/main/java/net/corda/client/jfx/utils/ReplayedList.java:6]: Kotlin forces you to specify the `override` keyword whilst Java doesn't.
Thus we can only make this cross-compatible via Java.

>>Re: wonder if it's worth being compatible with both 8 and 11 simultaneously

The intent is for this 1st stage migration to Java 11 to be 100% cross-compatible (single source, two targets: 8 & 11). Subsequent work with Java 11 will entail forking the code at some point.

File: client/rpc/build.gradle (hunk 3 above)
Comment (josecoll) [reply client/rpc/build.gradle:102]: Agree - this is an item that has been raised on several occasions but never put to bed properly (I am guessing we have some entanglement in our module dependencies that doesn't make it easy to remove).
I will raise a separate JIRA to investigate / resolve this properly (independent of this Java 11 migration PR).

File: node/src/main/kotlin/net/corda/node/serialization/kryo/DefaultKryoCustomizer.kt (hunk 4 above)
Comment (josecoll) [reply node/src/main/kotlin/net/corda/node/serialization/kryo/DefaultKryoCustomizer.kt:100]: Neither of these types are exported publicly:

import **sun.security.ec.ECPublicKeyImpl**:
```
module jdk.crypto.ec {
    provides java.security.Provider with sun.security.ec.SunEC;
}

```
import **sun.security.provider.certpath.X509CertPath**:

```
module java.base {
    ...
    exports sun.security.provider.certpath to java.naming;
    ...
}
```

Why are they being exposed for Serialization ?

File: node/src/main/kotlin/net/corda/node/utilities/JVMAgentUtil.kt (hunk 5 above)
Comment (josecoll) [reply node/src/main/kotlin/net/corda/node/utilities/JVMAgentUtil.kt:8]: This is a new utility used to provide a cross-compatible java 8 & 11 agent lookup mechanism.
See its usage in `NodeStartup` and `CheckpointDumper`.

File: samples/simm-valuation-demo/build.gradle (hunk 6 above)
Comment (josecoll) [reply samples/simm-valuation-demo/build.gradle:37]: I believe @chrisr3 attempted to fix this. 
Perhaps he can shed some light on how to remove this unnecessary boilerplate?

File: serialization/src/main/kotlin/net/corda/serialization/internal/carpenter/ClassCarpenter.kt (hunk 8 above)
Comment (josecoll) [reply serialization/src/main/kotlin/net/corda/serialization/internal/carpenter/ClassCarpenter.kt:349]: This fixes 4x Serialization unit tests failing with same root cause (ICCE):

```
:serialization:testnet.corda.serialization.internal.amqp.DeserializeNeedingCarpentryOfEnumsTest » singleEnum (0.016s)
java.lang.reflect.InvocationTargetException: (No message provided)Open stacktrace
Caused by: java.lang.IncompatibleClassChangeError: Method java.lang.Enum.valueOf(Ljava/lang/Class;Ljava/lang/String;)Ljava/lang/Enum; must be Methodref constant
```

Possible tightening of invokedynamic in JDK9:
http://hg.openjdk.java.net/jdk9/jdk9/hotspot/rev/ea45fb60fbbe

File: samples/simm-valuation-demo/contracts-states/build.gradle (hunk 7 above)
Comment (josecoll) [reply samples/simm-valuation-demo/contracts-states/build.gradle]: Will reset and test.

File: client/jfx/src/main/java/net/corda/client/jfx/utils/ReplayedList.java (hunk 2 above)
Comment (mikehearn) [reply client/jfx/src/main/java/net/corda/client/jfx/utils/ReplayedList.java:6]: Yeah, I get the intent. I just wonder if it'd be worth shrinking this PR and just ... doing it. <insert nike tick here>

Wouldn't it be a bit confusing for users of the platform if we supported 8 and 11 simultaneously? Can app devs use 11 features or not?

File: node/src/main/kotlin/net/corda/node/serialization/kryo/DefaultKryoCustomizer.kt (hunk 4 above)
Comment (mikehearn) [reply node/src/main/kotlin/net/corda/node/serialization/kryo/DefaultKryoCustomizer.kt:100]: Good question. I'm not sure. Maybe @rick-r3 remembers?

File: build.gradle [hunk 9]
```diff
@@ -201,9 +224,17 @@ allprojects {
...
         }
     }
-    sourceCompatibility = 1.8
-    targetCompatibility = 1.8
-
+    sourceCompatibility = VERSION_1_8
+    if (JavaVersion.current() == JavaVersion.VERSION_1_8)
+        targetCompatibility = VERSION_1_8
+    else
+        targetCompatibility = VERSION_11
```
Comment (rick-r3) [inline build.gradle:236]: Do the artefacts produced in any (external) way reflect their target version? e.g. in filenames.  I'm guessing not.  Maybe it would be something we would do when publishing to Artifactory etc.  Could get confusing though.

File: config/dev/log4j2.xml [hunk 10]
```diff
@@ -22,41 +22,12 @@
...
-                                } else {
-                                    result = null;
-                                }
-                                result;
-                                ]]>
-                            </Script>
-                            <PatternMatch key="WithMDC" pattern="%highlight{[%level{length=5}] %date{HH:mm:ssZ} [%t] %c{2}.%method - %msg %X%n%throwable{0}}{INFO=white,WARN=red,FATAL=bright red}"/>
-                        </ScriptPatternSelector>
-                    </PatternLayout>
+                    <PatternLayout pattern="%highlight{[%level{length=5}] %date{HH:mm:ssZ} [%t] %c{2}.%method - %msg%n%throwable{0}}{INFO=white,WARN=red,FATAL=bright red}"/>
```
Comment (rick-r3) [inline config/dev/log4j2.xml]: I vaguely remember discussing this.  This is because Nashorn is deprecated, right ?  What about the bits of Javascript a few lines above this?  Is it completely removed / now an additional JAR or just not available?

File: constants.properties [hunk 11]
```diff
@@ -3,7 +3,7 @@
 # their own projects. So don't get fancy with syntax!
 
 cordaVersion=4.3-SNAPSHOT
-gradlePluginsVersion=5.0.2
+gradlePluginsVersion=5.0.1-SNAPSHOT-quasar-jdk11
```
Comment (rick-r3) [inline constants.properties]: Remind me what is going on here?

File: core-deterministic/testing/src/test/java/net/corda/deterministic/CheatingSecurityProvider.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+import java.security.SecureRandom;
+import java.security.SecureRandomSpi;
+import java.security.Security;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Temporarily restore Sun's [SecureRandom] provider.
+ * This is ONLY for allowing us to generate test data, e.g. signatures.
```
Comment (rick-r3) [inline core-deterministic/testing/src/test/java/net/corda/deterministic/CheatingSecurityProvider.java:13]: Is this in Java so it can use the extra exports to see the Sun classes?  Worth a comment here.

File: core/src/main/kotlin/net/corda/core/internal/CordaUtils.kt [hunk 13]
```diff
@@ -41,6 +41,7 @@ fun checkMinimumPlatformVersion(minimumPlatformVersion: Int, requiredMinPlatform
     }
 }
 
+// JDK11: revisit (JDK 9+ uses different numbering scheme: see https://docs.oracle.com/javase/9/docs/api/java/lang/Runtime.Version.html)
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/internal/CordaUtils.kt:44]: When do we need to revisit ?

File: core/src/test/kotlin/net/corda/core/internal/X509EdDSAEngineTest.kt [hunk 14]
```diff
@@ -1,125 +0,0 @@
-package net.corda.core.internal
-
-import net.corda.core.crypto.Crypto
-import net.i2p.crypto.eddsa.EdDSAEngine
-import net.i2p.crypto.eddsa.EdDSAPublicKey
-import org.junit.Test
-import sun.security.util.BitArray
-import sun.security.util.ObjectIdentifier
```
Comment (rick-r3) [inline core/src/test/kotlin/net/corda/core/internal/X509EdDSAEngineTest.kt:8]: Could we restore this if we port to Java ?  Or did we decide these tests were duplicated elsewhere or something?

File: node/src/main/kotlin/net/corda/node/serialization/kryo/DefaultKryoCustomizer.kt (hunk 4 above)
Comment (rick-r3) [reply node/src/main/kotlin/net/corda/node/serialization/kryo/DefaultKryoCustomizer.kt:100]: Hmmm..... I wonder why we don't just register against the `PublicKey` and `PrivateKey` interfaces?

I'm fairly certain we could `addDefaultSerializer(PublicKey::class.java, publicKeySerializer)` and same for private.

File: node/src/test/kotlin/net/corda/node/services/network/DBNetworkParametersStorageTest.kt [hunk 15]
```diff
@@ -98,24 +96,11 @@ class DBNetworkParametersStorageTest {
...
     }
 
-    private fun interceptConsoleOutput(block: () -> Unit): List<String> {
-        val oldOut = System.out
-        val out = ByteOutputStream()
-        System.setOut(PrintStream(out))
-        block()
-        System.setOut(oldOut)
-        return out.bytes.inputStream().bufferedReader().lines().toList()
-    }
```
Comment (rick-r3) [inline node/src/test/kotlin/net/corda/node/services/network/DBNetworkParametersStorageTest.kt:117]: What doesn't work here?  Is `setOut` gone or something ?  Or is it the `ByteOutputStream` that looks like it was accidentally imported from the wrong place?

File: samples/simm-valuation-demo/contracts-states/build.gradle (hunk 7 above)
Comment (rick-r3) [reply samples/simm-valuation-demo/contracts-states/build.gradle]: and?

File: serialization-deterministic/build.gradle [hunk 16]
```diff
@@ -8,6 +10,9 @@ apply plugin: 'idea'
 
 evaluationDependsOn(":serialization")
 
+// required by DJVM and Avian JVM (for running inside the SGX enclave) which only supports Java 8.
+//targetCompatibility = VERSION_1_8
```
Comment (rick-r3) [inline serialization-deterministic/build.gradle]: Is this supposed to be commented out?

File: serialization/build.gradle [hunk 17]
```diff
@@ -1,9 +1,14 @@
+import static org.gradle.api.JavaVersion.VERSION_1_8
+
 apply plugin: 'kotlin'
 apply plugin: 'net.corda.plugins.publish-utils'
 apply plugin: 'com.jfrog.artifactory'
 
 description 'Corda serialization'
 
+// required by DJVM and Avian JVM (for running inside the SGX enclave) which only supports Java 8.
+//targetCompatibility = VERSION_1_8
```
Comment (rick-r3) [inline serialization/build.gradle]: What about this one?

File: testing/node-driver/src/main/java/net/corda/testing/driver/SharedMemoryIncremental.java [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.lang.reflect.Field;
+import java.nio.MappedByteBuffer;
+import java.nio.channels.FileChannel;
+
+class SharedMemoryIncremental extends PortAllocation {
```
Comment (rick-r3) [inline testing/node-driver/src/main/java/net/corda/testing/driver/SharedMemoryIncremental.java:19]: Worth a comment about it being in Java.

**Review by rick-r3 — COMMENTED** [review]
First pass

File: build.gradle (hunk 9 above)
Comment (josecoll) [reply build.gradle:236]: > Do the artefacts produced in any (external) way reflect their target version? e.g. in filenames. I'm guessing not. Maybe it would be something we would do when publishing to Artifactory etc. Could get confusing though.

Yes - that is a good point.
Perhaps we should use the POM classifier to differentiate between Java 8 and 11 artifacts.

As per http://maven.apache.org/pom.html
```
classifier:
The classifier distinguishes artifacts that were built from the same POM but differ in content. It is some optional and arbitrary string that - if present - is appended to the artifact name just after the version number.
As a motivation for this element, consider for example a project that offers an artifact targeting JRE 1.5 but at the same time also an artifact that still supports JRE 1.4. The first artifact could be equipped with the classifier jdk15 and the second one with jdk14 such that clients can choose which one to use.

File: client/jfx/src/main/java/net/corda/client/jfx/utils/ReplayedList.java (hunk 2 above)
Comment (josecoll) [reply client/jfx/src/main/java/net/corda/client/jfx/utils/ReplayedList.java:6]: That is an interesting question .... to which I think the answer is "maybe".
the Public API's (core module) and Serialization are still built to generate 100% Java 8 bytecode.
but that shouldn't stop CorDapps using JDK11 features right?
Time to adapt a sample standalone CorDapp to validate this.

File: config/dev/log4j2.xml (hunk 10 above)
Comment (josecoll) [reply config/dev/log4j2.xml]: Hmm ... this was one of those horrible run-time class visibility exception scenarios.
It manifested itself upon executing the _example-code:deployNodes_ gradle task:

```
[:example-code:deployNodes] 2019-06-05 00:28:33,803 Execution worker for ':' INFO Cannot initialize scripting support because this JRE does not support it. java.lang.NoClassDefFoundError: javax/script/ScriptEngineManager
[00:28:34]
[:example-code:deployNodes] 	at org.apache.logging.log4j.core.script.ScriptManager.<init>(ScriptManager.java:69)
```
See details in:
https://stackoverflow.com/questions/53714010/log4j2-slf4j-and-java-11

Both the above and https://issues.apache.org/jira/browse/LOG4J2-2488 suggest we need to execute the deploy command with `--add-modules java.scripting`. 
However, if you look at the module descriptor for `javax.script` you can see this package is already being exported:

```
module java.scripting {
    exports javax.script;

    uses javax.script.ScriptEngineFactory;
}
```

File: serialization/build.gradle (hunk 17 above)
Comment (josecoll) [reply serialization/build.gradle]: OK - I can see what is happening here now.
This module has a _**testCompile**_ dependency on **_:node-driver_** (which is compiled to Jaava 11 bytecode), thus giving the following compile time exception:

```
* What went wrong:
Could not determine the dependencies of task ':serialization:test'.
> Could not resolve all task dependencies for configuration ':serialization:testRuntimeClasspath'.
   > Could not resolve project :test-common.
     Required by:
         project :serialization
      > Unable to find a matching variant of project :test-common:
          - Variant 'apiElements' capability net.corda:test-common:4.3-SNAPSHOT:
              - Incompatible attributes:
                  - Required org.gradle.jvm.version '8' and found incompatible value '11'.
                  - Required org.gradle.usage 'java-runtime' and found incompatible value 'java-api-jars'.
              - Other attributes:
                  - Found org.gradle.category 'library' but wasn't required.
                  - Required org.gradle.dependency.bundling 'external' and found compatible value 'external'.
                  - Required org.jetbrains.kotlin.platform.type 'jvm' and found compatible value 'jvm'.
          - Variant 'runtimeElements' capability net.corda:test-common:4.3-SNAPSHOT:
              - Incompatible attribute:
                  - Required org.gradle.jvm.version '8' and found incompatible value '11'.
              - Other attributes:
                  - Found org.gradle.category 'library' but wasn't required.
                  - Required org.gradle.dependency.bundling 'external' and found compatible value 'external'.
                  - Required org.gradle.usage 'java-runtime' and found compatible value 'java-runtime-jars'.
                  - Required org.jetbrains.kotlin.platform.type 'jvm' and found compatible value 'jvm'.
```
The solution to this is to split out the _serialization_ module tests into a separate module (as @blsemo has recently done for the _core_ module *and* using a different package name to avoid the split packages problem when we come to adding module descriptors to these modules).

File: serialization-deterministic/build.gradle (hunk 16 above)
Comment (josecoll) [reply serialization-deterministic/build.gradle]: see my comment below.

File: testing/node-driver/src/main/java/net/corda/testing/driver/SharedMemoryIncremental.java (hunk 18 above)
Comment (josecoll) [reply testing/node-driver/src/main/java/net/corda/testing/driver/SharedMemoryIncremental.java:19]: Yes - although @roastario mentioned to me the other day that they (the infrastructure team???) are ripping out this approach and reverting to a non-intrusive parallelisation mechanism (which I think is the correct way of solving the problem properly)

File: client/rpc/build.gradle (hunk 3 above)
Comment (josecoll) [reply client/rpc/build.gradle:102]: @blsemo would you care to share you thoughts (as Modularity WG lead) on how we might want to approach this?

File: config/dev/log4j2.xml (hunk 10 above)
Comment (josecoll) [reply config/dev/log4j2.xml]: I have re-instated the original configuration as it no longer appears to be causing a problem (when just built locally). Lets see what a full TC build / test cycle thinks.

File: constants.properties (hunk 11 above)
Comment (josecoll) [reply constants.properties]: Aha! .... I need to update this to the latest gradle snapshot which should automatically handle the correct version of quasar according to the setting of `quasarVersion` and `quasarClassifier` in the main project gradle file.

File: core/src/main/kotlin/net/corda/core/internal/CordaUtils.kt (hunk 13 above)
Comment (josecoll) [reply core/src/main/kotlin/net/corda/core/internal/CordaUtils.kt:44]: I guess when we obtain and prescribe a specific version of Java 11.
The current patched Azul Zulu version we are using is:

```
colljos@ci-agent-101l:~$ java -version
openjdk version "11.0.3" 2019-04-16 LTS
OpenJDK Runtime Environment Zulu11.31+16-SA (build 11.0.3+7-LTS)
OpenJDK 64-Bit Server VM Zulu11.31+16-SA (build 11.0.3+7-LTS, mixed mode)
```

File: node/src/test/kotlin/net/corda/node/services/network/DBNetworkParametersStorageTest.kt (hunk 15 above)
Comment (josecoll) [reply node/src/test/kotlin/net/corda/node/services/network/DBNetworkParametersStorageTest.kt:117]: The test was using an internal class called: `com.sun.xml.internal.messaging.saaj.util.ByteOutputStream` to capture console output as a means of testing the correctness of an internal method (that should return an Exception but returns `null`).

File: core/src/test/kotlin/net/corda/core/internal/X509EdDSAEngineTest.kt (hunk 14 above)
Comment (josecoll) [reply core/src/test/kotlin/net/corda/core/internal/X509EdDSAEngineTest.kt:8]: How bizarre .... I ported these tests to Java ... on the Enterprise feature branch (doh!!!).
Will add to this feature branch.

File: serialization/build.gradle (hunk 17 above)
Comment (josecoll) [reply serialization/build.gradle]: I have created a separate JIRA to track this work (as it will be merged directly into the mainline OS and ENT branches): https://r3-cev.atlassian.net/browse/CORDA-3206

File: samples/simm-valuation-demo/contracts-states/build.gradle (hunk 7 above)
Comment (josecoll) [reply samples/simm-valuation-demo/contracts-states/build.gradle]: Reset and re-tested SIMM Valuation demo. All worked fine.

File: constants.properties (hunk 11 above)
Comment (josecoll) [reply constants.properties]: Repointed to latest published version of Gradle Plugins (which contains quasar version and classifier parsing): https://ci-artifactory.corda.r3cev.com/artifactory/corda-releases/net/corda/plugins/quasar-utils/5.0.2/

File: webserver/src/main/java/CordaWebserverCaplet.java [hunk 19]
```diff
@@ -174,8 +174,8 @@ private void augmentClasspath(List<Path> classpath, File dir) {
 
     private static void checkJavaVersion() {
         String version = System.getProperty("java.version");
-        if (version == null || !version.startsWith("1.8")) {
-            System.err.printf("Error: Unsupported Java version %s; currently only version 1.8 is supported.\n", version);
+        if (version == null || Arrays.asList("1.8", "11").stream().noneMatch(version::startsWith)) {
```
Comment (shamsasari) [inline webserver/src/main/java/CordaWebserverCaplet.java]: You can simplify this with `Stream.of("1.8", "11")`

File: samples/simm-valuation-demo/contracts-states/build.gradle (hunk 7 above)
Comment (josecoll) [reply samples/simm-valuation-demo/contracts-states/build.gradle]: OK - So the automated integration test is failing:

```
Caused by: net.corda.core.CordaRuntimeException: class "net.corda.vega.analytics.OGSIMMAnalyticsEngine"'s signer information does not match signer information of other classes in the same package
Caused by: net.corda.core.CordaRuntimeException: java.lang.SecurityException: class "net.corda.vega.analytics.OGSIMMAnalyticsEngine"'s signer information does not match signer information of other classes in the same package

net.corda.vega.SimmValuationTest > runs SIMM valuation demo FAILED
    java.io.IOException at SimmValuationTest.kt:82
```

Will troubleshoot further.

File: build.gradle (hunk 9 above)
Comment (josecoll) [reply build.gradle:236]: @rick-r3 as discussed, lets do this as a separate follow-up PR.
I've created a JIRA story for this: https://r3-cev.atlassian.net/browse/CORDA-3224

File: core-deterministic/testing/src/test/java/net/corda/deterministic/CheatingSecurityProvider.java (hunk 12 above)
Comment (josecoll) [reply core-deterministic/testing/src/test/java/net/corda/deterministic/CheatingSecurityProvider.java:13]: I have added similar comments elsewhere for Java re-written code.

File: serialization/build.gradle (hunk 17 above)
Comment (josecoll) [reply serialization/build.gradle]: This is now completed and has been merged into this PR.

File: testing/node-driver/src/main/java/net/corda/testing/driver/SharedMemoryIncremental.java (hunk 18 above)
Comment (josecoll) [reply testing/node-driver/src/main/java/net/corda/testing/driver/SharedMemoryIncremental.java:19]: Added comment.

**Review by rick-r3 — APPROVED** [review]
Can you work with @r3domfox team to create a PR gate for JDK 11 compatibility ?  I guess up to them whether they run per PR or we have something like the Windows build, but the latter isn't good at keeping things green.  Create a JIRA to track if we don't have one.

**Comment by josecoll** [comment]
Yes - I created a JIRA to create a JDK11 compilation-only check gate: https://r3-cev.atlassian.net/browse/DEVOPS-1367
I propose we only run *tests* overnight as part of a snapshot build.
